File: backend-ts/src/domain/budget-ledger.ts
export type BudgetItemType = "attraction" | "hotel" | "meal" | "transport" | "other";
export type AmountBasis = "group_total" | "per_person";

export interface BudgetLedgerItem {
  id: string;
  type: BudgetItemType;
  day_index: number | null;
  day_end_index: number | null;
  name: string;
  amount: number | null;
  amount_basis: AmountBasis;
  traveler_count: number;
  per_person_amount: number | null;
  calculation_summary: string;
  unit_amount: number | null;
  room_count: number | null;
  nights: number | null;
  origin: "itinerary" | "user";
  price_source: "unavailable" | "estimated" | "live" | "user";
  price_provider: string;
  linked_item_id: string;
  entity_source: string;
  source_url: string;
  price_checked_at: string;
  note: string;
  user_locked: boolean;
  deleted: boolean;
}

export interface BudgetTotals {
  total_attractions: number;
  total_hotels: number;
  total_meals: number;
  total_transportation: number;
  total_inter_city_transport: number;
  total_other: number;
  total: number;
}

const TYPE_ORDER: Record<BudgetItemType, number> = {
  attraction: 0,
  hotel: 1,
  meal: 2,
  transport: 3,
  other: 4,
};

function record(value: unknown): Record<string, unknown> | null {
  return value && typeof value === "object" && !Array.isArray(value)
    ? value as Record<string, unknown>
    : null;
}

function amount(value: unknown): number {
  const numeric = Number(value ?? 0);
  return Number.isFinite(numeric) ? Math.round(Math.max(0, numeric) * 100) / 100 : 0;
}

function count(value: unknown, fallback = 1): number {
  const numeric = Number(value);
  return Number.isFinite(numeric) ? Math.max(1, Math.trunc(numeric)) : fallback;
}

function perPerson(value: number | null, travelers: number): number | null {
  return value === null ? null : Math.round(value / count(travelers) * 100) / 100;
}

function baseItem(values: Partial<BudgetLedgerItem> & Pick<BudgetLedgerItem, "id" | "type" | "name">): BudgetLedgerItem {
  return {
    id: values.id,
    type: values.type,
    day_index: values.day_index ?? null,
    day_end_index: values.day_end_index ?? null,
    name: values.name,
    amount: values.amount ?? null,
    amount_basis: values.amount_basis ?? "group_total",
    traveler_count: values.traveler_count ?? 1,
    per_person_amount: values.per_person_amount ?? null,
    calculation_summary: values.calculation_summary ?? "",
    unit_amount: values.unit_amount ?? null,
    room_count: values.room_count ?? null,
    nights: values.nights ?? null,
    origin: values.origin ?? "itinerary",
    price_source: values.price_source ?? "unavailable",
    price_provider: values.price_provider ?? "",
    linked_item_id: values.linked_item_id ?? "",
    entity_source: values.entity_source ?? "",
    source_url: values.source_url ?? "",
    price_checked_at: values.price_checked_at ?? "",
    note: values.note ?? "",
    user_locked: values.user_locked ?? false,
    deleted: values.deleted ?? false,
  };
}

export function toGroupTotal(
  value: number | null,
  basis: AmountBasis,
  travelerCount: number,
): number | null {
  if (value === null) return null;
  const normalized = amount(value) * (basis === "per_person" ? count(travelerCount) : 1);
  return Math.round(normalized * 100) / 100;
}

function deriveHotelItems(days: unknown[], travelers: number, rooms: number): BudgetLedgerItem[] {
  const items: BudgetLedgerItem[] = [];
  let group: {
    identity: string;
    hotel: Record<string, unknown>;
    start: number;
    end: number;
    nights: number;
  } | null = null;
  const dayIndices = days.map((day, index) => count(record(day)?.day_index ?? index, index || 1) - (index === 0 ? 1 : 0));
  const finalDayIndex = dayIndices.length ? Math.max(...dayIndices) : 0;

  const flush = () => {
    if (!group) return;
    const current = group;
    const priceStatus = String(current.hotel.price_status ?? "unavailable");
    const rawUnit = amount(current.hotel.estimated_cost);
    const unit = priceStatus === "unavailable" || rawUnit === 0 ? null : rawUnit;
    const total = unit === null ? null : Math.round(unit * rooms * current.nights * 100) / 100;
    items.push(baseItem({
      id: `itinerary:hotel:${current.start}`,
      type: "hotel",
      day_index: current.start,
      day_end_index: Math.min(current.end + 1, finalDayIndex),
      name: String(current.hotel.name || "未命名酒店"),
      amount: total,
      amount_basis: "group_total",
      traveler_count: travelers,
      per_person_amount: perPerson(total, travelers),
      unit_amount: unit,
      room_count: rooms,
      nights: current.nights,
      calculation_summary: "room_night",
      price_source: priceStatus === "live" || priceStatus === "estimated" ? priceStatus : "unavailable",
      price_provider: String(current.hotel.price_source ?? ""),
      linked_item_id: String(current.hotel.source_hotel_id ?? ""),
      entity_source: String(current.hotel.source ?? ""),
      source_url: String(current.hotel.source_url ?? ""),
      price_checked_at: String(current.hotel.price_checked_at ?? ""),
    }));
    group = null;
  };

  days.forEach((rawDay, position) => {
    const day = record(rawDay) ?? {};
    const dayIndex = Number.isInteger(Number(day.day_index)) ? Number(day.day_index) : position;
    const hotel = record(day.hotel);
    if (!hotel) {
      flush();
      return;
    }
    const identity = JSON.stringify([
      hotel.source ?? "",
      hotel.source_hotel_id ?? "",
      hotel.name ?? "",
      amount(hotel.estimated_cost),
    ]);
    if (group && group.identity === identity && group.end + 1 === dayIndex) {
      group.end = dayIndex;
      group.nights += 1;
      return;
    }
    flush();
    group = { identity, hotel, start: dayIndex, end: dayIndex, nights: 1 };
  });
  flush();
  return items;
}

export function deriveBudgetItems(
  result: unknown,
  travelerCount?: number,
  roomCount?: number,
): BudgetLedgerItem[] {
  const plan = record(record(result)?.data);
  if (!plan) return [];
  const travelers = count(travelerCount ?? plan.traveler_count);
  const rooms = count(roomCount ?? plan.room_count, Math.ceil(travelers / 2));
  const days = Array.isArray(plan.days) ? plan.days : [];
  const budget = record(plan.budget) ?? {};
  const transportTotal = amount(budget.total_transportation);
  const transportDays = days.filter((rawDay) => String(record(rawDay)?.transportation ?? "").trim());
  const transportShare = transportDays.length ? Math.round(transportTotal / transportDays.length * 100) / 100 : 0;
  let transportRemaining = transportTotal;
  const items: BudgetLedgerItem[] = [];

  days.forEach((rawDay, dayPosition) => {
    const day = record(rawDay) ?? {};
    const dayIndex = Number.isInteger(Number(day.day_index)) ? Number(day.day_index) : dayPosition;
    const attractions = Array.isArray(day.attractions) ? day.attractions : [];
    attractions.forEach((rawAttraction, itemPosition) => {
      const attraction = record(rawAttraction) ?? {};
      const linkedId = String(attraction.id || `${dayIndex}-${itemPosition}`);
      const unit = amount(attraction.ticket_price);
      const total = Math.round(unit * travelers * 100) / 100;
      items.push(baseItem({
        id: `itinerary:attraction:${linkedId}`,
        type: "attraction",
        day_index: dayIndex,
        name: String(attraction.name || "未命名景点"),
        amount: total,
        amount_basis: "per_person",
        traveler_count: travelers,
        per_person_amount: unit,
        unit_amount: unit,
        calculation_summary: "per_person",
        price_source: "estimated",
        linked_item_id: linkedId,
        entity_source: attraction.poi_id ? "amap" : "",
      }));
    });

    const meals = Array.isArray(day.meals) ? day.meals : [];
    meals.forEach((rawMeal, itemPosition) => {
      const meal = record(rawMeal) ?? {};
      const linkedId = String(meal.id || `${dayIndex}-${itemPosition}`);
      const unit = amount(meal.estimated_cost);
      const total = Math.round(unit * travelers * 100) / 100;
      items.push(baseItem({
        id: `itinerary:meal:${linkedId}`,
        type: "meal",
        day_index: dayIndex,
        name: String(meal.name || "未命名餐饮"),
        amount: total,
        amount_basis: "per_person",
        traveler_count: travelers,
        per_person_amount: unit,
        unit_amount: unit,
        calculation_summary: "per_person",
        price_source: "estimated",
        linked_item_id: linkedId,
      }));
    });

    const transportation = String(day.transportation ?? "").trim();
    if (transportation) {
      const isLast = transportDays.at(-1) === rawDay;
      const share = isLast ? transportRemaining : Math.min(transportShare, transportRemaining);
      transportRemaining = Math.round(Math.max(0, transportRemaining - share) * 100) / 100;
      const total = transportTotal > 0 ? Math.round(share * 100) / 100 : null;
      items.push(baseItem({
        id: `itinerary:transport:${dayIndex}`,
        type: "transport",
        day_index: dayIndex,
        name: transportation,
        amount: total,
        traveler_count: travelers,
        per_person_amount: perPerson(total, travelers),
        calculation_summary: "shared_total",
        price_source: total === null ? "unavailable" : "estimated",
        linked_item_id: `day:${dayIndex}:transport`,
      }));
    }
  });

  items.push(...deriveHotelItems(days, travelers, rooms));
  return items;
}

function validSavedItem(value: unknown): value is BudgetLedgerItem {
  const item = record(value);
  return Boolean(item && typeof item.id === "string" && item.id && typeof item.name === "string"
    && ["attraction", "hotel", "meal", "transport", "other"].includes(String(item.type)));
}

export function syncBudgetItems(
  result: unknown,
  savedItems: unknown[] | null,
  travelerCount?: number,
  roomCount?: number,
): BudgetLedgerItem[] {
  const plan = record(record(result)?.data);
  const travelers = count(travelerCount ?? plan?.traveler_count);
  const saved = (savedItems ?? []).filter(validSavedItem).map((item) => ({
    ...structuredClone(item),
    price_provider: String(item.price_provider ?? ""),
  }));
  const byId = new Map(saved.map((item) => [item.id, item]));
  const derived = deriveBudgetItems(result, travelers, roomCount);
  const derivedIds = new Set(derived.map((item) => item.id));
  const merged = derived.map((item) => {
    const previous = byId.get(item.id);
    if (!previous?.user_locked) return item;
    previous.traveler_count = travelers;
    previous.per_person_amount = perPerson(previous.amount, travelers);
    previous.day_end_index = item.day_end_index;
    previous.room_count = item.room_count;
    previous.nights = item.nights;
    if (previous.price_source !== "user") {
      previous.unit_amount = item.unit_amount;
      previous.calculation_summary = item.calculation_summary;
    }
    return previous;
  });
  for (const item of saved) {
    if (!derivedIds.has(item.id) && (item.origin === "user" || item.user_locked)) {
      item.traveler_count = travelers;
      item.per_person_amount = perPerson(item.amount, travelers);
      merged.push(item);
    }
  }
  return merged.sort((left, right) => Number(left.deleted) - Number(right.deleted)
    || (left.day_index ?? 10_000) - (right.day_index ?? 10_000)
    || TYPE_ORDER[left.type] - TYPE_ORDER[right.type]
    || left.id.localeCompare(right.id));
}

export function calculateBudgetTotals(items: BudgetLedgerItem[]): BudgetTotals {
  const totals: BudgetTotals = {
    total_attractions: 0,
    total_hotels: 0,
    total_meals: 0,
    total_transportation: 0,
    total_inter_city_transport: 0,
    total_other: 0,
    total: 0,
  };
  const target: Record<BudgetItemType, keyof BudgetTotals> = {
    attraction: "total_attractions",
    hotel: "total_hotels",
    meal: "total_meals",
    transport: "total_transportation",
    other: "total_other",
  };
  for (const item of items) {
    if (item.deleted || item.amount === null) continue;
    const value = amount(item.amount);
    totals[target[item.type]] += value;
    totals.total += value;
  }
  for (const key of Object.keys(totals) as Array<keyof BudgetTotals>) {
    totals[key] = Math.round(totals[key] * 100) / 100;
  }
  return totals;
}

export function budgetLedgerResponse(planId: string, task: {
  result: unknown;
  request_payload: Record<string, unknown> | null;
  budget_items: unknown[] | null;
}): Record<string, unknown> {
  const plan = record(record(task.result)?.data) ?? {};
  const request = task.request_payload ?? {};
  const travelers = count(request.traveler_count ?? plan.traveler_count);
  const rooms = count(request.room_count ?? plan.room_count, Math.ceil(travelers / 2));
  const items = syncBudgetItems(task.result, task.budget_items, travelers, rooms);
  task.budget_items = items;
  const totals = calculateBudgetTotals(items);
  plan.budget = { ...(record(plan.budget) ?? {}), ...totals };
  const perPersonTotals = Object.fromEntries(
    Object.entries(totals).map(([key, value]) => [key, Math.round(value / travelers * 100) / 100]),
  );
  const rawLimit = Number(request.budget_amount ?? plan.budget_amount);
  const validLimit = Number.isFinite(rawLimit) && rawLimit > 0;
  const budgetLimit = validLimit
    ? Math.round(rawLimit * (String(request.budget_basis ?? plan.budget_basis) === "per_person" ? travelers : 1) * 100) / 100
    : null;
  const pending = items.filter((item) => !item.deleted && item.amount === null);
  const pendingBuffer = pending.reduce((sum, item) => {
    if (item.type === "hotel") return sum + 300 * (item.room_count ?? 1) * (item.nights ?? 1);
    const perTraveler = item.type === "meal" ? 80 : item.type === "transport" ? 50 : 100;
    return sum + perTraveler * travelers;
  }, 0);
  const quotedTotal = totals.total;
  const projectedTotal = Math.round((quotedTotal + pendingBuffer) * 100) / 100;
  return {
    plan_id: planId,
    items,
    totals,
    per_person_totals: perPersonTotals,
    traveler_count: travelers,
    room_count: rooms,
    pending_count: pending.length,
    budget_limit: budgetLimit,
    quoted_total: quotedTotal,
    over_budget_amount: budgetLimit === null ? 0 : Math.round(Math.max(0, quotedTotal - budgetLimit) * 100) / 100,
    pending_buffer: pendingBuffer,
    projected_total: projectedTotal,
    projected_over_budget_amount: budgetLimit === null ? 0 : Math.round(Math.max(0, projectedTotal - budgetLimit) * 100) / 100,
    adjustment_applied: Boolean(plan.budget_adjustment_applied),
    adjustment_note: String(plan.budget_adjustment_note ?? ""),
  };
}

File: backend.bak/app/services/budget_ledger.py
"""Persistent, user-editable budget ledger derived from a trip plan."""

from __future__ import annotations

from typing import Any, Iterable, Literal

from ..models.schemas import BudgetLedgerItem


_TYPE_ORDER = {
    "attraction": 0,
    "hotel": 1,
    "meal": 2,
    "transport": 3,
    "other": 4,
}


def _get(value: Any, key: str, default: Any = None) -> Any:
    if isinstance(value, dict):
        return value.get(key, default)
    return getattr(value, key, default)


def _plan(result: Any) -> Any:
    return _get(result, "data")


def _amount(value: Any) -> float:
    try:
        number = float(value or 0)
    except (TypeError, ValueError):
        return 0
    return round(max(0, number), 2)


def _count(value: Any, default: int = 1) -> int:
    try:
        return max(1, int(value))
    except (TypeError, ValueError):
        return default


def to_group_total(
    amount: float | None,
    amount_basis: Literal["group_total", "per_person"],
    traveler_count: int,
) -> float | None:
    """将用户录入金额换算成台账唯一存储口径：全体人员合计金额。"""
    if amount is None:
        return None
    normalized = _amount(amount)
    if amount_basis == "per_person":
        normalized *= _count(traveler_count)
    return round(normalized, 2)


def _per_person(amount: float | None, traveler_count: int) -> float | None:
    if amount is None:
        return None
    return round(_amount(amount) / _count(traveler_count), 2)


def _default_item(**values: Any) -> BudgetLedgerItem:
    return BudgetLedgerItem(
        origin="itinerary",
        user_locked=False,
        deleted=False,
        note="",
        **values,
    )


def _hotel_items(
    days: list[Any], travelers: int, rooms: int
) -> list[BudgetLedgerItem]:
    items: list[BudgetLedgerItem] = []
    group: dict[str, Any] | None = None

    def flush() -> None:
        nonlocal group
        if group is None:
            return
        hotel = group["hotel"]
        nights = group["nights"]
        price_status = str(
            _get(hotel, "price_status", "unavailable") or "unavailable"
        )
        unit_amount = (
            None
            if price_status == "unavailable"
            else _amount(_get(hotel, "estimated_cost", 0))
        )
        if unit_amount == 0:
            unit_amount = None
        total_amount = (
            None
            if unit_amount is None
            else round(unit_amount * rooms * nights, 2)
        )
        start_index = group["start_index"]
        final_day_index = max(
            (int(_get(day, "day_index", position)) for position, day in enumerate(days)),
            default=group["end_index"],
        )
        checkout_index = min(group["end_index"] + 1, final_day_index)
        items.append(_default_item(
            id=f"itinerary:hotel:{start_index}",
            type="hotel",
            day_index=start_index,
            day_end_index=checkout_index,
            name=str(_get(hotel, "name", "") or "未命名酒店"),
            amount=total_amount,
            amount_basis="group_total",
            traveler_count=travelers,
            per_person_amount=_per_person(total_amount, travelers),
            unit_amount=unit_amount,
            room_count=rooms,
            nights=nights,
            calculation_summary="room_night",
            price_source=(
                price_status
                if price_status in {"live", "estimated"}
                else "unavailable"
            ),
            linked_item_id=str(_get(hotel, "source_hotel_id", "") or ""),
            entity_source=str(_get(hotel, "source", "") or ""),
            source_url=str(_get(hotel, "source_url", "") or ""),
            price_checked_at=str(_get(hotel, "price_checked_at", "") or ""),
        ))
        group = None

    for day_position, day in enumerate(days):
        day_index = int(_get(day, "day_index", day_position))
        hotel = _get(day, "hotel")
        if hotel is None:
            flush()
            continue
        identity = (
            str(_get(hotel, "source", "") or ""),
            str(_get(hotel, "source_hotel_id", "") or ""),
            str(_get(hotel, "name", "") or ""),
            _amount(_get(hotel, "estimated_cost", 0)),
        )
        can_extend = (
            group is not None
            and group["identity"] == identity
            and group["end_index"] + 1 == day_index
        )
        if can_extend:
            group["end_index"] = day_index
            group["nights"] += 1
            continue
        flush()
        group = {
            "identity": identity,
            "hotel": hotel,
            "start_index": day_index,
            "end_index": day_index,
            "nights": 1,
        }
    flush()
    return items


def derive_budget_items(
    result: Any,
    traveler_count: int | None = None,
    room_count: int | None = None,
) -> list[BudgetLedgerItem]:
    """Build canonical group-total ledger rows from the current itinerary."""
    plan = _plan(result)
    if plan is None:
        return []

    travelers = _count(traveler_count or _get(plan, "traveler_count", 1))
    rooms = _count(
        room_count or _get(plan, "room_count", (travelers + 1) // 2),
        (travelers + 1) // 2,
    )
    days = list(_get(plan, "days", []) or [])
    budget = _get(plan, "budget")
    transport_total = _amount(_get(budget, "total_transportation", 0))
    transport_days = [
        day for day in days if str(_get(day, "transportation", "") or "").strip()
    ]
    transport_share = (
        round(transport_total / len(transport_days), 2) if transport_days else 0
    )
    transport_remaining = transport_total
    items: list[BudgetLedgerItem] = []

    for day_position, day in enumerate(days):
        day_index = int(_get(day, "day_index", day_position))
        for item_position, attraction in enumerate(_get(day, "attractions", []) or []):
            linked_id = str(
                _get(attraction, "id", "") or f"{day_index}-{item_position}"
            )
            unit_amount = _amount(_get(attraction, "ticket_price", 0))
            total_amount = round(unit_amount * travelers, 2)
            items.append(_default_item(
                id=f"itinerary:attraction:{linked_id}",
                type="attraction",
                day_index=day_index,
                name=str(_get(attraction, "name", "") or "未命名景点"),
                amount=total_amount,
                amount_basis="per_person",
                traveler_count=travelers,
                per_person_amount=unit_amount,
                unit_amount=unit_amount,
                calculation_summary="per_person",
                price_source="estimated",
                linked_item_id=linked_id,
                entity_source="amap" if _get(attraction, "poi_id", "") else "",
            ))

        for item_position, meal in enumerate(_get(day, "meals", []) or []):
            linked_id = str(_get(meal, "id", "") or f"{day_index}-{item_position}")
            unit_amount = _amount(_get(meal, "estimated_cost", 0))
            total_amount = round(unit_amount * travelers, 2)
            items.append(_default_item(
                id=f"itinerary:meal:{linked_id}",
                type="meal",
                day_index=day_index,
                name=str(_get(meal, "name", "") or "未命名餐饮"),
                amount=total_amount,
                amount_basis="per_person",
                traveler_count=travelers,
                per_person_amount=unit_amount,
                unit_amount=unit_amount,
                calculation_summary="per_person",
                price_source="estimated",
                linked_item_id=linked_id,
            ))

        transportation = str(_get(day, "transportation", "") or "").strip()
        if transportation:
            is_last = day is transport_days[-1] if transport_days else False
            share = transport_remaining if is_last else min(
                transport_share, transport_remaining
            )
            transport_remaining = round(max(0, transport_remaining - share), 2)
            total_amount = round(share, 2) if transport_total > 0 else None
            items.append(_default_item(
                id=f"itinerary:transport:{day_index}",
                type="transport",
                day_index=day_index,
                name=transportation,
                amount=total_amount,
                amount_basis="group_total",
                traveler_count=travelers,
                per_person_amount=_per_person(total_amount, travelers),
                calculation_summary="shared_total",
                price_source="estimated" if total_amount is not None else "unavailable",
                linked_item_id=f"day:{day_index}:transport",
            ))

    items.extend(_hotel_items(days, travelers, rooms))
    return items


def sync_budget_items(
    result: Any,
    saved_items: Iterable[dict[str, Any] | BudgetLedgerItem] | None,
    traveler_count: int | None = None,
    room_count: int | None = None,
) -> list[BudgetLedgerItem]:
    """Refresh generated rows while preserving every user-owned override."""
    saved: list[BudgetLedgerItem] = []
    for raw in saved_items or []:
        try:
            saved.append(BudgetLedgerItem.model_validate(raw))
        except (TypeError, ValueError):
            continue

    plan = _plan(result)
    travelers = _count(traveler_count or _get(plan, "traveler_count", 1))
    saved_by_id = {item.id: item for item in saved}
    derived = derive_budget_items(result, travelers, room_count)
    merged: list[BudgetLedgerItem] = []
    derived_ids: set[str] = set()
    for item in derived:
        derived_ids.add(item.id)
        previous = saved_by_id.get(item.id)
        if previous and previous.user_locked:
            previous.traveler_count = travelers
            previous.per_person_amount = _per_person(previous.amount, travelers)
            previous.day_end_index = item.day_end_index
            previous.room_count = item.room_count
            previous.nights = item.nights
            if previous.price_source != "user":
                previous.unit_amount = item.unit_amount
                previous.calculation_summary = item.calculation_summary
            selected = previous
        else:
            selected = item
        merged.append(selected)

    for item in saved:
        if item.id not in derived_ids and (item.origin == "user" or item.user_locked):
            item.traveler_count = travelers
            item.per_person_amount = _per_person(item.amount, travelers)
            merged.append(item)

    return sorted(
        merged,
        key=lambda item: (
            item.deleted,
            item.day_index if item.day_index is not None else 10_000,
            _TYPE_ORDER.get(item.type, 99),
            item.id,
        ),
    )


def calculate_budget_totals(items: Iterable[BudgetLedgerItem]) -> dict[str, float]:
    totals = {
        "total_attractions": 0.0,
        "total_hotels": 0.0,
        "total_meals": 0.0,
        "total_transportation": 0.0,
        "total_inter_city_transport": 0.0,
        "total_other": 0.0,
        "total": 0.0,
    }
    targets = {
        "attraction": "total_attractions",
        "hotel": "total_hotels",
        "meal": "total_meals",
        "transport": "total_transportation",
        "other": "total_other",
    }
    for item in items:
        if item.deleted or item.amount is None:
            continue
        amount = _amount(item.amount)
        totals[targets[item.type]] += amount
        totals["total"] += amount
    return {key: round(value, 2) for key, value in totals.items()}


def calculate_per_person_totals(
    totals: dict[str, float], traveler_count: int
) -> dict[str, float]:
    travelers = _count(traveler_count)
    return {
        key: round(_amount(value) / travelers, 2)
        for key, value in totals.items()
    }


def apply_budget_totals(result: Any, totals: dict[str, float]) -> None:
    plan = _plan(result)
    if plan is None:
        return
    budget = _get(plan, "budget")
    if isinstance(plan, dict):
        if not isinstance(budget, dict):
            budget = {}
            plan["budget"] = budget
        budget.update(totals)
        return
    if budget is None:
        from ..models.schemas import Budget

        budget = Budget(**totals)
        plan.budget = budget
        return
    for key, value in totals.items():
        setattr(budget, key, value)
